core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,776 @@
+	return res;
+}
+
+static TEE_Result stmm_enter_open_session(struct tee_ta_session *s,
+					  struct tee_ta_param *param __unused,
+					  TEE_ErrorOrigin *eo)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+
+	if (spc->is_initializing) {
+		/* stmm is initialized in sec_part_init_session() */
+		*eo = TEE_ORIGIN_TEE;
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stmm_enter_invoke_cmd(struct tee_ta_session *s,
+					uint32_t cmd,
+					struct tee_ta_param *param,
+					TEE_ErrorOrigin *eo __unused)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+	TEE_Result res = TEE_SUCCESS;
+	vaddr_t ns_buf_base = 0;
+	unsigned int ns_buf_size;
+
+	if (cmd != PTA_STMM_COMMUNICATE)
+		return TEE_ERROR_NOT_IMPLEMENTED;
+
+	ns_buf_size = param->u[0].mem.size;
+	if (ns_buf_size > spc->ns_comm_buf_size)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = stmm_map_ns_buf(spc, param, &ns_buf_base);
+	if (res)
+		return res;
+
+	spc->regs.x[0] = FFA_MSG_SEND_DIRECT_REQ_64;
+	spc->regs.x[1] = (stmm_pta_id << 16) | stmm_id;
+	spc->regs.x[2] = FFA_PARAM_MBZ;
+	spc->regs.x[3] = spc->ns_comm_buf_addr;
+	spc->regs.x[4] = ns_buf_size;
+	spc->regs.x[5] = 0;
+	spc->regs.x[6] = 0;
+	spc->regs.x[7] = 0;
+
+	tee_ta_push_current_session(s);
+
+	memcpy((void *)(spc->ns_comm_buf_addr), (void *)ns_buf_base,
+	       ns_buf_size);
+
+	res = sec_part_enter_user_mode(spc);
+
+	/* Error handling TBD */
+	if (!res)
+		param->u[1].val.a = spc->regs.x[1];
+
+	memcpy((void *)ns_buf_base, (void *)(spc->ns_comm_buf_addr),
+	       ns_buf_size);
+
+	/*
+	 * Clear out the parameter mappings added with tee_mmu_map_param()
+	 * above.
+	 */
+	tee_mmu_clean_param(&spc->uctx);
+
+	tee_ta_pop_current_session();
+
+	return res;
+}
+
+static void stmm_enter_close_session(struct tee_ta_session *s __unused)
+{
+}
+
+static void sec_part_dump_state(struct tee_ta_ctx *ctx)
+{
+	user_mode_ctx_print_mappings(to_user_mode_ctx(ctx));
+}
+
+static uint32_t sec_part_get_instance_id(struct tee_ta_ctx *ctx)
+{
+	return to_sec_part_ctx(ctx)->uctx.vm_info.asid;
+}
+
+static void sec_part_ctx_destroy(struct tee_ta_ctx *ctx)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(ctx);
+
+	tee_pager_rem_um_areas(&spc->uctx);
+	vm_info_final(&spc->uctx);
+	free(spc);
+}
+
+static int sp_svc_set_mem_attr(vaddr_t va, unsigned int nr_pages, uint32_t perm)
+{
+	TEE_Result res;
+	struct tee_ta_session *sess = NULL;
+	struct sec_part_ctx *spc = NULL;
+	uint32_t prot;
+
+	if (va == 0 || nr_pages == 0)
+		return SP_RET_INVALID_PARAM;
+
+	res = tee_ta_get_current_session(&sess);
+	if (res != TEE_SUCCESS)
+		return SP_RET_DENIED;
+
+	spc = to_sec_part_ctx(sess->ctx);
+
+	prot = 0;
+	if ((perm & SP_MEM_ATTR_ACCESS_MASK) == SP_MEM_ATTR_ACCESS_RW)
+		prot |= TEE_MATTR_URW | TEE_MATTR_PRW;
+	else if ((perm & SP_MEM_ATTR_ACCESS_MASK) == SP_MEM_ATTR_ACCESS_RO)
+		prot |= TEE_MATTR_UR | TEE_MATTR_PR;
+
+	if ((perm & SP_MEM_ATTR_EXEC_MASK) == SP_MEM_ATTR_EXEC) {
```

Comment:
I'd prefer refusing permissions that we cannot or will not support.

Reply:
Fair enough, I'll return an error in these cases

Reply:
Here's what happens here, StrMM before setting the flags of a page, reads the current state and ORs it with the new flags it wants to set. 
The reason for masking out the writable flag if the execute bit is set, is because we return a all flags (RWX) on our get function. Looking into OP-TEE there's a vm_get_flags(), but no vm_get_attrs(). How about adding a helper for getting thr attributes and returning the proper page attributes back to StMM?

Reply:
> Looking into OP-TEE there's a vm_get_flags(), but no vm_get_attrs(). How about adding a helper for getting the attributes and returning the them back to StMM?

Please do.